### PR TITLE
Add support for setting capabilities on the app binary

### LINKED ISSUE #1246: Support setting capabilities on the app binary
Linux has a notion of [capabilities](https://manpages.debian.org/unstable/manpages/capabilities.7.en.html), which is essentially a *token* allowing a certain privileged operation.

E.g. `CAP_BPF` allows loading ebpf programs for an otherwise unprivileged user.

The way to leverage capabilities with Docker is two fold:
 * at build time, use `setcap` tool to set file capabilities on a binary;
 * at run time, [request](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities) *matching* capabilities via `--cap-add` option.

We are doing ebpf in go and we'd like to take advantage of ko's fast build times and convenience. We need a way to set custom capabilities on the app binary.

It looks like currently it is not supported, as far as I can tell from docs and [implementation](https://github.com/ko-build/ko/blob/main/pkg/build/gobuild.go#L529).

Would you be open for such contribution?

### PR DESCRIPTION
**Pull request by mejedi**
This patchset adds `linux_capabilities` under `builds` [section](https://ko.build/configuration/#overriding-go-build-settings) in `.ko.yaml`. Ex:
```yaml
builds:
- id: my-app
  dir: ./cmd
  main: ./myapp
  linux_capabilities: bpf perfmon net_admin
```

A bit of trivia on capabilities: unless running the app as root user, Docker's `--cap-add` alone is insufficient.
Requested capabilities are AND-combined with capabilities granted to the app binary (intentionally simplified, see `man capabilities` for the full discourse).

A running program has multiple sets of capabilities. The most important ones are `effective` (used by the kernel for permission checks) and `permitted` (a "stash" of capabilities an app can promote to effective). Likewise, file capabilities capture multiple sets of capabilities, `permitted` being the most important. 

```
linux_capabilities: bpf perfmon net_admin
```

The config above sets `permitted` capabilities on the app binary to `bpf`, `perfmon`, and `net_admin`. When the app is launched, this set is AND-combined with capabilities requested via `--cap-add` and the result becomes the running app's `permitted` capabilities. The app should verify if it got all required capabilities and promote `permitted` capabilities to `effective`.

File capabilities have a bit that tells the kernel to automatically promote `permitted` to `effective`. The downside is that the program will fail to start with generic `EPERM` error if some capabilities weren't granted. In order to access this feature, and to make transition from `Dockerfile` easier, we also support the full `setcap` syntax in `linux_capabilities`, e.g.:

```
linux_capabilities: bpf,perfmon,net_admin=ep
```

Fixes #1246

### CODE REVIEW

**Review by imjasonh — COMMENTED** [review]
This looks really good! Thanks for the exhaustive test cases, that makes reviewing this for correctness a lot easier.

Just a few code style things, nothing major.

File: pkg/caps/doc.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package caps implements a subset of Linux capabilities handling
+// relevant in the context of authoring container images.
+package caps
```
[the remark this answers is not in the document]
Comment (mejedi) [reply pkg/caps/doc.go:17]: Done.

File: pkg/caps/new_file_caps_test.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+func TestNewFileCaps(t *testing.T) {
+	type testCase struct {
+		args     []string
+		res      *FileCaps
+		mustFail bool
+	}
+	tests := []testCase{
+		testCase{},
```
[the remark this answers is not in the document]
Comment (mejedi) [reply pkg/caps/new_file_caps_test.go:30]: Done.

File: pkg/caps/caps_test.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+package caps
+
+import (
+	"encoding/base64"
+	"fmt"
+	"testing"
+)
+
+func TestParseBasic(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mejedi) [reply pkg/caps/caps_test.go:23]: Done.

**Comment by mejedi** [comment]
@imjasonh 
> Just a few code style things, nothing major.

Thank you for taking a look! I've implemented  suggestions and fixed issues reported by linters.

**Review by imjasonh — APPROVED** [review]

**Pull request merged**